### PR TITLE
Add `FileContextKind` `trait` for more safety

### PR DESCRIPTION
**Pull request by TimeEngineer**

### CODE REVIEW

File: winfsp_wrs/src/callback.rs [hunk 1]
```diff
@@ -12,11 +14,75 @@ use crate::{
         SECURITY_INFORMATION, SIZE_T, UINT32, UINT64, ULONG,
     },
     CleanupFlags, CreateFileInfo, CreateOptions, DirInfo, FileAccessRights, FileAttributes,
-    FileInfo, PSecurityDescriptor, SecurityDescriptor, VolumeInfo,
+    FileContextMode, FileInfo, PSecurityDescriptor, SecurityDescriptor, VolumeInfo,
 };
 
+/// Implement only if necessary at your own risk
+pub trait FileContextKind {
+    fn mode() -> FileContextMode;
```
Comment (touilleMan) [inline winfsp_wrs/src/callback.rs:22]: Can we use an [associated constant](https://doc.rust-lang.org/reference/items/associated-items.html#associated-constants) instead ?

```rust
trait FileContextKind {
  const MODE: FileContextMode;
...
}
```

File: winfsp_wrs/src/callback.rs [hunk 2]
```diff
@@ -12,11 +14,75 @@ use crate::{
...
+    /// # Safety
+    ///
+    /// Retrieve the data from winfsp's `FileContext`
+    /// This is called in winfsp-wrs before calling FileSystemContext::close
+    unsafe fn access_for_close(raw: PVOID) -> Self;
+}
+
+impl<T> FileContextKind for Arc<T> {
+    fn mode() -> FileContextMode {
+        FileContextMode::Descriptor
```
Comment (touilleMan) [inline winfsp_wrs/src/callback.rs:42]: `FileContextMode::Full` (i.e. `UmFileContextIsFullContext=1`) is missing in `FileContextMode`

You should add it to `FileContextMode`: this way it's still possible to use this mode (but it's up to the user to implement `impl<T> FileContextKind for MyFullContextStuff`, which seems okay as we talked earlier)

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: winfsp_wrs/src/callback.rs (hunk 1 above)
Comment (TimeEngineer) [reply winfsp_wrs/src/callback.rs:22]: Yes, nice suggestion

File: winfsp_wrs/src/callback.rs (hunk 2 above)
Comment (TimeEngineer) [reply winfsp_wrs/src/callback.rs:42]: Sure 👍

**Review by touilleMan — APPROVED** [review]
:1st_place_medal:

**Pull request merged**